src/sparseml/transformers/nm_training_arguments.py

```diff
@@ -0,0 +1,96 @@
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+NMTrainingArguments class extends the original HF TrainingArguments class
+to include additional arguments
+"""
+
+from dataclasses import dataclass, field
+from typing import Optional
+
+from transformers import TrainingArguments
+
+
+@dataclass
+class NMTrainingArguments(TrainingArguments):
+    # Add support to total batch size arguments
+    total_train_batch_size: Optional[int] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Total batch size for training."
+                "If defined, is used to compute batch size per "
+                "GPU/TPU core/CPU."
+            )
+        },
+    )
+    total_eval_batch_size: Optional[int] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Total batch size for evaluation."
+                "If defined, is used to compute batch size "
+                "per GPU/TPU core/CPU."
+            )
+        },
+    )
+
+    # Override per device batch size arguments to remove default values
+    per_device_train_batch_size: Optional[int] = field(
+        default=None, metadata={"help": "Batch size per GPU/TPU core/CPU for training."}
+    )
+    per_device_eval_batch_size: Optional[int] = field(
+        default=None,
+        metadata={"help": "Batch size per GPU/TPU core/CPU for evaluation."},
+    )
+
+    #
+    def __post_init__(self):
+        # Execute post_init from TrainingArguments class
+        super().__post_init__()
+
+        # Raise error if both total and per device batch sizes are defined
+        if self.total_train_batch_size and self.per_device_train_batch_size:
+            raise ValueError(
+                (
+                    "Cannot specify both total_train_batch_size and"
+                    " per_device_train_batch_size."
+                )
+            )
+        if self.total_eval_batch_size and self.per_device_eval_batch_size:
+            raise ValueError(
+                (
+                    "Cannot specify both total_eval_batch_size and"
+                    " per_device_eval_batch_size."
+                )
+            )
+
+        # If neither total nor per device batch sizes are defined,
+        # set total batch sizes as default
+        if not self.total_train_batch_size and not self.per_device_train_batch_size:
+            self.total_train_batch_size = 8
+        if not self.total_eval_batch_size and not self.per_device_eval_batch_size:
+            self.total_eval_batch_size = 8
+
+        # If total batch size is defined at this point,
+        # per device batch size needs to be computed
+        if self.total_train_batch_size:
+            self.per_device_train_batch_size = self.total_train_batch_size // max(
```

Comment:
Would it be better if we use this total_train_batch_size to dictate both the per_device_train_batch_size and the gradient_accumulation_steps (so the "steps" here mean backward steps)---this way we make sure that this total batch size, if specified, is consistent even between runs with different num gpus? We could start with an accumulation step of 1, calculating the corresponding per device batch size; if that fails then we double the accumulation step (and so on).

Reply:
Hi Tuan. I fail to see how gradient_accumulation_steps can be set "consistently" from the total batch size alone. In my understanding, the accumulation batch size will be total_train_batch_size * gradient_accumulation_steps. On the other hand, per_device_train_batch_size = total_batch_size / nGPUs, regardless of the accumulation steps.

Reply:
Hi Alex, my thinking was that we could interpret "total_train_batch_size" as your "accumulation batch size" above, which I think is consistent with the base trainer code here: https://github.com/neuralmagic/transformers/blob/master/src/transformers/trainer.py#L1085
in which the args.train_batch_size I think is computed from per_device_batch_size and num gpus as here: https://github.com/neuralmagic/transformers/blob/master/src/transformers/training_args.py#L699

So taking all together, we would have total_train_batch_size is a func of per_device_batch_size, gradient_accumulation_steps, and num gpus. We know num gpus, and then I'm thinking we can adjust the other two params for the batch to fit gpu memory.

Reply:
I think I understand that now. However, how would we check if it the batch fits into GPU memory? I would rather allow the user to specify either total_train_batch_size or per_device_train_batch_size (together with gradient_accumulation_steps) and compute the other from the expression in L1085 in trainer.py above. Does that address your concern?

Reply:
Please review the latest version. This one should have the total batch size take the accumulation steps into consideration